File: python/tests/parity/producer.py
import asyncio
from datetime import datetime
from uuid import uuid4

from bullmq import Queue
from tests.parity.utils import getParityBackendOptions, logEvent, readDefinitions


async def createJobs():
    backendOptions = getParityBackendOptions()
    definitions = readDefinitions()

    logEvent("ready")
    for definition in definitions:
        queue = Queue(definition.get("id", ""), backendOptions)

        testSecret = str(uuid4())
        for i in range(int(definition.get("job", {}).get("count"))):
            jobName = f"job-{i}"
            jobSecret = str(uuid4())
            await queue.add(
                jobName,
                {"test_secret": testSecret, "job_secret": jobSecret},
                {"delay": definition.get("job", {}).get("delay")},
            )

            logEvent(
                "job-created",
                {
                    "timestamp": int(datetime.now().timestamp() * 1000),
                    "test_id": str(definition.get("id")),
                    "job_name": jobName,
                    "test_secret": testSecret,
                    "job_secret": jobSecret,
                },
            )

        await queue.close()


if __name__ == "__main__":
    asyncio.run(createJobs())

File: rust/tests/parity/src/producer.rs
use core::time::Duration;

use bullmq::{options::RedisConnectionOptions, JobOptions, Queue, QueueOptions};
use uuid::Uuid;

use crate::{
    models::{JobData, ParityEventData},
    utils::{log_event, read_producer_definitions},
};

pub async fn create_jobs(port: u16) -> bullmq::Result<()> {
    let definitions = read_producer_definitions()?;

    let mut connection = RedisConnectionOptions::default();
    connection.url = format!("redis://localhost:{port}");

    log_event("ready".to_string(), None);

    for definition in definitions {
        let options = QueueOptions::default().connection(connection.clone());
        let queue = Queue::with_options(&definition.id, options).await?;

        let test_secret = Uuid::new_v4().to_string();
        for i in 0..definition.job.count {
            let job_name = format!("job-{i}");
            let job_secret = Uuid::new_v4().to_string();

            let job_data = JobData {
                job_secret,
                test_secret: test_secret.clone(),
            };

            let options = if definition.job.delay > 0 {
                JobOptions::default().delay(Duration::from_millis(definition.job.delay))
            } else {
                JobOptions::default()
            };

            queue
                .add(&job_name, job_data.clone())
                .options(options)
                .await?;

            log_event(
                "job-created".to_string(),
                ParityEventData::new_some(definition.id.clone(), job_name, job_data),
            );
        }
    }

    Ok(())
}
